Source/DafnyCore/DafnyPrelude.bpl

```diff
@@ -505,6 +505,21 @@ function AtLayer<A>([LayerType]A, LayerType): A;
 axiom (forall<A> f : [LayerType]A, ly : LayerType :: { AtLayer(f,ly) } AtLayer(f,ly) == f[ly]);
 axiom (forall<A> f : [LayerType]A, ly : LayerType :: { AtLayer(f,$LS(ly)) } AtLayer(f,$LS(ly)) == AtLayer(f,ly));
 
+// ---------------------------------------------------------------
+// -- Opaqueness encoding ----------------------------------------
+// ---------------------------------------------------------------
+
+function CanRevealFuel(LayerType, bool) : LayerType;
+
+axiom (forall $ly: LayerType :: 
+  { CanRevealFuel($LS($ly), true) }
+  CanRevealFuel($LS($ly), true) == $LS(CanRevealFuel($ly, true)));
+
+axiom (forall $ly: LayerType, b: bool ::
+  { CanRevealFuel(AsFuelBottom($ly), b) }
+  CanRevealFuel(AsFuelBottom($ly), b)
+  == $LZ);
```

Comment:
Well done in figuring out the problem and suggesting a solution. I don't think we ever equate anything with `$LZ` elsewhere. Because of that, my hunch is that it's better to write the axiom as

``` boogie
axiom (forall $ly: LayerType, b: bool :: 
  { CanRevealFuel(AsFuelBottom($ly), b) } 
  CanRevealFuel(AsFuelBottom($ly), b) == AsFuelBottom($ly));
```

In fact, even more, since the first axiom (LL 514-516) says that `CanRevealFuel` commutes with `$LS`, provided the second parameter is known to equal `true`, I think it seems appropriate to do the analogous thing for the second axiom. That is, I suggest writing the second axiom as

``` boogie
axiom (forall $ly: LayerType, b: bool :: 
  { CanRevealFuel(AsFuelBottom($ly), b) } 
  CanRevealFuel(AsFuelBottom($ly), b) == AsFuelBottom(CanRevealFuel($ly, b)));
```